### DISCUSSION TITLE
Facial recognition - switching to newer model

### DISCUSSION THREAD

**Discussion by maltokyo** [immich-app/immich#5028] [General]
With the newest release, v1.8.6 - I switched to `antelopev2` from `buffalo_l` - now if I go to rerun the recognition job from scratch, it tells me I will lose all tagged faces.  What is your workflow to move to a better/larger model?  After tagging/labeling 100s of faces, I do not want to have to do it all again, and with that being said, I want to use the best model possible.  Will this be needed every time we upgrade the model?  (I would have hoped it keeps existing labels, but perhaps recognises even more faces)

**Comment by thomasverelst** [comment]
Aside from your question, it's not clear whether the `antelopev2` is actually better. I can't find any benchmark numbers comparing the two, but from this https://github.com/deepinsight/insightface/issues/1820 it looks like the "r100 glint360k model" (which seems antelopev2, https://github.com/deepinsight/insightface/blob/master/python-package/README.md) performs worse than the `buffalo_l` model because the latter is trained on more data with a better training schedule. Seems like there is a reason that insightface does not promote the `antelopev2` model and prefers the `buffalo_l`.

Glint360k (of antelopev2) is 360k identities and 17M images from what I can find.
WebFace600k seems to have 600k identities but I can't find the number of images. Seems like there exist multiple versions of WebFace, including variants with 2M identities.

**Reply by mertalev** [reply]
Thank you for looking into this! I wasn't able to find any documentation on antelopev2. I took a look at the links you mentioned.

Given they have several versions of WebFace600k R50 models, and buffalo_l is from a later iteration not included in the table of generic recognition models, could it be a similar case for antelopev2? It isn't clear that the reported metrics for the Glint360k R100 model are actually representative of antelopev2. If they are, though, then I think this would make buffalo_l outclass antelopev2.

Either way, it's very strange that we need to guess like this. Pending more info, I'm hesitant to recommend antelopev2 given we don't have any evidence that it performs better than buffalo_l, and possible evidence of the opposite.

**Reply by thomasverelst** [reply]
Seems like there was indeed an original antelope, added on May 2021 (https://github.com/deepinsight/insightface/commit/0fdf664d68b6558665d67fdea6cb4af7c852f41e) which was changed to antelopev2 in July 2021 (https://github.com/deepinsight/insightface/commit/8420187ed95c467b8d566df412794b03a261343d). Can't find accuracy numbers, should probably just try to run the benchmark myself 😅

**Reply by aviv926** [reply]
WebFace600K(600K identities) is same as WebFace12M(12 million images), a subset of WebFace42M(cleaned WebFace260M).

source:
https://github.com/deepinsight/insightface/issues/1770#issuecomment-927131443

**Comment by mertalev** [comment]
To answer your question: the reason you can't change the model and keep the relations is because we don't currently have separate jobs for model inference and clustering. 

Clustering involves comparing model outputs for one face to outputs for other faces. The current algorithm clusters on-the-fly after receiving the model outputs for a face. As a result, all of these outputs have to come from the same model for it to work correctly. And since the same job is also responsible for generating these outputs, replacing the model also means clustering from scratch. 

In the future, we hope to separate these into two jobs that can be run independently. Switching models wouldn't necessarily mean discarding clusters, and re-clustering would be faster as it wouldn't require running the model on all images again.

**Reply by maltokyo** [reply]
Thank you, very helpful

**Reply by JvdMaat** [reply]
With 1.93 now having separate facial recognition (clustering) from facial detection, can we now switch algorithm without starting from scratch?

**Reply by mertalev** [reply]
Not at the moment. There's no guarantee that face detection will produce the same faces and bounding boxes. More work would be needed to handle this.

**Reply by OffsetMonkey538** [reply]
Any chance this will be on the roadmap?

**Reply by liamrotheram** [reply]
Keen to keep bumping this!

**Comment by Vyerni** [comment]
I switched to the newer model, though silly me didn't take notes of what exactly I had numbers of what before. 

Now I'm starting to question whether it is an actual improvement or not.

**Comment by davidpan** [comment]
antelopev2 is not better...... -_-!

**Reply by Vyerni** [reply]
Right... 

Might need to duplicate the machine it's on and switch back on the other machine, to compare... Means going through all the faces, again 🤦‍♂️

**Comment by tsemachh** [comment]
Assuming we would like to switch model (Or just run facial job with tweaked settings)
- Iterate all named persons find an asset where they appear and has only 1 face - store the asset id + person data
- If we can't find then store the asset id + the face box data + person data
- run again the facial process
- go other the original saved data , restore the person data based on new recognition
Not sure how easy to make this process automatic , maybe there's better way
